src/Engine/gramambular2/reading_grid.h

```diff
@@ -0,0 +1,229 @@
+      // top unigram. This is used for soft-override such as from a suggestion.
+      // The node with the override value will be likely to favored by a walk,
+      // but it does not prevent other nodes from prevailing, which would be the
+      // case if kOverrideValueWithHighScore was used.
+      kOverrideValueStealingScoreFromTopUnigram
+    };
+
+    Node(std::string reading, size_t readingLength,
+         std::vector<LanguageModel::Unigram> unigrams)
+        : reading_(std::move(reading)),
+          readingLength_(readingLength),
+          unigrams_(std::move(unigrams)),
+          unigramIter_(unigrams_.begin()),
+          overrideType_(OverrideType::kNone) {}
+
+    [[nodiscard]] const std::string& reading() const { return reading_; }
+
+    [[nodiscard]] size_t readingLength() const { return readingLength_; }
+
+    [[nodiscard]] const std::vector<LanguageModel::Unigram>& unigrams() const {
+      return unigrams_;
+    }
+
+    [[nodiscard]] std::string value() const;
+
+    [[nodiscard]] double score() const;
+
+    [[nodiscard]] bool isOverridden() const;
+
+    void reset();
+
+    bool selectOverrideUnigram(const std::string& value, OverrideType type);
+
+    // A sufficiently high score to cause the walk to go through an overriding
+    // node. Although this can be 0, setting it to a positive value has the
+    // desirable side effect that it reduces the competition of "free-floating"
+    // multiple-character phrases. For example, if the user override for
+    // reading "a b c" is "A B c", using the uppercase as the overriding note,
+    // now the standalone c may have to compete with a phrase with reading "bc",
+    // which in some pathological cases may actually cause the shortest path to
+    // be A->bc, especially when A and B use the zero overriding score, as they
+    // leave "c" alone to compete with "bc", and whether the path A-B is favored
+    // now solely depends on that competition. A positive value favors the route
+    // A->B, which gives "c" a better chance.
+    static constexpr double kOverridingScore = 42;
+
+   protected:
+    const std::string reading_;
+    const size_t readingLength_;
+    const std::vector<LanguageModel::Unigram> unigrams_;
+    std::vector<LanguageModel::Unigram>::const_iterator unigramIter_;
+    OverrideType overrideType_;
+  };
+
+  using NodePtr = std::shared_ptr<Node>;
+
+  struct WalkResult {
+    std::vector<NodePtr> nodes;
+    size_t vertices;
+    size_t edges;
+    uint64_t elapsedMicroseconds;
+
+    std::vector<std::string> valuesAsStrings();
+    std::vector<std::string> readingsAsStrings();
+  };
+
+  WalkResult walk();
+
+  // Returns all candidate values at the location. If spans are not empty and
+  // loc is at the end of the spans, (loc - 1) is used, so that the caller does
+  // not have to care about this boundary condition.
+  std::vector<std::string> candidatesAt(size_t loc);
```

Comment:
Good point. I guess we could return `(reading, value)` instead of `(spanningLength, value)` as reading disambiguates better.

Practically, though, we've been doing overrides solely based on values for this long already, how often is that really a problem? If that's really a problem I'm open to extending the API as a future task.